plutus-ledger-api/src/PlutusLedgerApi/Common/Versions.hs

```diff
@@ -78,6 +80,32 @@ builtinsIntroducedIn = Map.fromList [
           ])
   ]
 
+-- | An array "associating" every plutus language version to its protocol version which first introduced that plutus version.
+-- OPTIMIZE: use SmallArray
+-- OPTIMIZE: or even use unboxed ByteArray/PrimArray (assuming minor protocolversions don't introduce language versions).
+langArray :: Array LedgerPlutusVersion ProtocolVersion
+langArray = array (minBound, maxBound)
+    [ (PlutusV1, alonzoPV)
+    , (PlutusV2, vasilPV)
+    , (PlutusV3, changPV)
+    ]
+
+-- | Query the protocol version that a specific ledger plutus version was first introduced in.
+-- 'Introduction' in this context means the enablement/allowance of scripts of that language version to be executed on-chain.
+languageIntroducedIn :: LedgerPlutusVersion -> ProtocolVersion
+languageIntroducedIn = (Array.!) langArray
```

Comment:
This is partial!! `Maybe`, please! Or if you want to have a function here, then maybe just write the function directly and skip the array?

Reply:
> This is partial!! `Maybe`, please! Or if you want to have a function here, then maybe just write the function directly and skip the array?

Is it?  I worried about that too, but the array in indexed by the `LedgerPlutusVersion` type, so you can't give `!` an out-of-bounds index, can you? Or am I missing something?

Reply:
Michael has a point. Although this code does not have a problem, using a function instead of a list to populate the array is safer because it provides static typecheck that the pattern match is exhaustive (total function)

Reply:
Yeah, this function can't fail, but `!` is partial. My problem with this sort of usually not the actual usage where it's introduced, it's that it gives us an opportunity to screw up in future accidentally. It's usually not when you write the partial function that you get the error, it's when you later modify it...